nemo_aligner/models/mm/stable_diffusion/megatron_sd_draftp_model.py

```diff
@@ -0,0 +1,381 @@
+            dp=parallel_state.get_data_parallel_world_size(),
+        )
+        self.onload_adam_states()
+
+    def onload_adam_states(self):
+        if self.distributed_adam_offload_manager is not None:
+            # load back onto GPU
+            self.distributed_adam_offload_manager.__exit__(None, None, None)
+
+        self.distributed_adam_offload_manager = None
+
+    def finish_training(self):
+        """no need to offload adam states here
+        """
+
+    def get_forward_output_and_loss_func(self, validation_step=False):
+        def fwd_output_and_loss_func(data_iterator, model):
+
+            batch = next(data_iterator)
+            batch_size = len(batch)
+            torch.cuda.manual_seed(torch.distributed.get_rank())
+            torch.manual_seed(torch.distributed.get_rank())
+            latents = torch.randn(
+                [
+                    batch_size,
+                    self.in_channels,
+                    self.height // self.downsampling_factor,
+                    self.width // self.downsampling_factor,
+                ],
+                generator=None,
+            ).to(torch.cuda.current_device())
+
+            output_tensor_draft_p, epsilons_draft_p, epsilons_init = self.generate(batch, latents)
+
+            # in this nemo version the model and autocast dtypes are not synced
+            # so we need to explicitly cast it
+            if not parallel_state.is_pipeline_last_stage():
+                output_tensor_draft_p = output_tensor_draft_p.to(dtype=self.autocast_dtype)
+
+            def loss_func(output_tensor_draft_p):
+                # Loss per micro batch (ub).
+                if self.cfg.kl_coeff == 0.0:
+                    kl_penalty = torch.tensor([0.0]).to(torch.cuda.current_device())
+                else:
+                    kl_penalty = calculate_gaussian_kl_penalty_shared_var(epsilons_draft_p, epsilons_init).mean()
+                rewards = self.reward_model.get_reward(
+                    output_tensor_draft_p.permute(0, 2, 3, 1), batch
+                )  # (ub, H, W, C) -> (ub, C, H, W)
+                loss = -rewards.mean() + kl_penalty * self.cfg.kl_coeff
+
+                reduced_loss = average_losses_across_data_parallel_group([loss])
+                reduced_kl_penalty = average_losses_across_data_parallel_group([kl_penalty])
+
+                return (
+                    loss,
+                    {"loss": reduced_loss, "kl_penalty": reduced_kl_penalty},
+                )
+
+            return output_tensor_draft_p, loss_func
+
+        return fwd_output_and_loss_func
+
+    def get_loss_and_metrics(self, batch, forward_only=False):
+
+        data_iter = get_iterator_k_split_list(batch, get_num_microbatches())
+
+        fwd_bwd_function = get_forward_backward_func()
+        losses_reduced_per_micro_batch = fwd_bwd_function(
+            forward_step_func=self.get_forward_output_and_loss_func(forward_only),
+            data_iterator=[data_iter],
```

Comment:
`get_iterator_k_split_list` chunks your batch into a list of lists, can't we just use data_iter instead of `[data_iter]` here?

Reply:
right right